### ISSUE TITLE
Reverse mode can truncate & corrupt files

### ISSUE THREAD

**Issue by benrubson** [vgough/encfs#383]
Hello,

Here is what I can reproduce, at least on FreeBSD.

Create a Fuse EncFS in reverse mode :
```
plain=/tmp/encfs_test/plain
ciphr=/tmp/encfs_test/ciphr
rm -rf $plain $ciphr
mkdir -p $plain $ciphr
encfs -fv --extpass="echo test" --standard $plain $ciphr --reverse
```
Create a file in the plain folder :
```
# echo 12345 >$plain/file
```
Stat its cipher version :
```
# stat $ciphr/*
3976265621 710805 -rw------- 1 root wheel 4294967295 6 "Aug 12 13:22:46 2017" "Aug 12 13:22:46 2017" "Aug 12 13:22:46 2017" "Jan  1 00:59:59 1970" 4096 9 0 /tmp/encfs_test/ciphr/pmxCPcKfIKM6-gmpYsLsdpVk
3976265621 710802 -rw------- 1 root wheel 4294967295 1264 "Aug 12 13:17:35 2017" "Aug 12 13:17:35 2017" "Aug 12 13:17:35 2017" "Jan  1 00:59:59 1970" 4096 9 0 /tmp/encfs_test/ciphr/XE7ixEhABTaxY24utXpTD06y
---
VER doing statfs of /tmp/encfs_test/plain [encfs.cpp:658]
VER readdir on /tmp/encfs_test/plain/ [encfs.cpp:211]
VER created FileNode for /tmp/encfs_test/plain/file [DirNode.cpp:644]
VER op: getattr : /tmp/encfs_test/plain/file [encfs.cpp:128]
VER created FileNode for /tmp/encfs_test/plain/.encfs6.xml [DirNode.cpp:644]
VER op: getattr : /tmp/encfs_test/plain/.encfs6.xml [encfs.cpp:128]
```
Grow the plain file :
```
# echo 12345 >>$plain/file
# echo 12345 >>$plain/file
# echo 12345 >>$plain/file
```
Stat again :
```
# stat $ciphr/*
3976265621 710805 -rw------- 1 root wheel 4294967295 6 "Aug 12 13:22:46 2017" "Aug 12 13:23:15 2017" "Aug 12 13:23:15 2017" "Jan  1 00:59:59 1970" 4096 9 0 /tmp/encfs_test/ciphr/pmxCPcKfIKM6-gmpYsLsdpVk
3976265621 710802 -rw------- 1 root wheel 4294967295 1264 "Aug 12 13:17:35 2017" "Aug 12 13:17:35 2017" "Aug 12 13:17:35 2017" "Jan  1 00:59:59 1970" 4096 9 0 /tmp/encfs_test/ciphr/XE7ixEhABTaxY24utXpTD06y
---
VER doing statfs of /tmp/encfs_test/plain [encfs.cpp:658]
VER readdir on /tmp/encfs_test/plain/ [encfs.cpp:211]
VER created FileNode for /tmp/encfs_test/plain/file [DirNode.cpp:644]
VER op: getattr : /tmp/encfs_test/plain/file [encfs.cpp:128]
VER created FileNode for /tmp/encfs_test/plain/.encfs6.xml [DirNode.cpp:644]
VER op: getattr : /tmp/encfs_test/plain/.encfs6.xml [encfs.cpp:128]
```
_As you can see only the modification date is updated, not the size, certainly because FreeBSD does not support `entry_timeout` / `attr_timeout` (#315) yet_.

Control the plain file :
```
# ls -l $plain/file
-rw-------  1 root  wheel  24 12 Aug 13:23 /tmp/encfs_test/plain/file
# cat $plain/file
12345
12345
12345
12345
```
Read the cipher file :
```
# dd if=$ciphr/pmxCPcKfIKM6-gmpYsLsdpVk of=/dev/null
0+1 records in
0+1 records out
6 bytes transferred in 0.000260 secs (23081 bytes/sec)
---
VER created FileNode for /tmp/encfs_test/plain/file [DirNode.cpp:644]
VER open call, requestWrite = 0 [RawFileIO.cpp:116]
VER open file with flags 0, result = 4 [RawFileIO.cpp:135]
VER encfs_open for /tmp/encfs_test/plain/file, flags 0 [encfs.cpp:544]
VER op: getattr : /tmp/encfs_test/plain/file [encfs.cpp:128]
VER op: read : /tmp/encfs_test/plain/file [encfs.cpp:128]
VER relaying request to base class: offset=0, dataLen=6 [CipherFileIO.cpp:460]
VER streamRead(data, 24, IV) [CipherFileIO.cpp:331]
VER op: truncate : /tmp/encfs_test/plain/file [encfs.cpp:128]
VER streamRead(data, 24, IV) [CipherFileIO.cpp:331]
VER Called streamWrite [CipherFileIO.cpp:396]
ERROR Assert failed: true == canWrite [RawFileIO.cpp:211]
ERROR withFileNode: error caught in truncate: true == canWrite [encfs.cpp:148]
VER op: truncate : /tmp/encfs_test/plain/file [encfs.cpp:128]
VER op: getattr : /tmp/encfs_test/plain/file [encfs.cpp:128]
```
Control the plain file :
```
# ls -l $plain/file
-rw-------  1 root  wheel  6 12 Aug 13:23 /tmp/encfs_test/plain/file
# cat $plain/file
12345
```

So the plain file has been truncated after having read its cipher version !
**Really dangerous data loss !**

I've exposed the truncate risk in the FreeBSD `entry_timeout` / `attr_timeout` [support request](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=218101).

However, I wonder if Linux could also be impacted, even if it supports these parameters.
There is certainly a race condition where this could occur.

Isn't reverse mode expected to be read-only ?

Thank you,

Ben

**Comment by benrubson** [comment]
OK, this is a side effect of #301.

**Comment by benrubson** [comment]
OK, solved by #386 !
